sdk/core/azure_core_amqp/src/sender.rs

```diff
@@ -139,57 +145,57 @@ pub mod builders {
                 options: Default::default(),
             }
         }
-        #[allow(dead_code)]
-        pub fn with_sender_settle_mode(mut self, sender_settle_mode: SenderSettleMode) -> Self {
+
+        pub fn with_sender_settle_mode(&mut self, sender_settle_mode: SenderSettleMode) -> &mut Self {
             self.options.sender_settle_mode = Some(sender_settle_mode);
             self
         }
-        #[allow(dead_code)]
+
         pub fn with_receiver_settle_mode(
-            mut self,
+            &mut self,
             receiver_settle_mode: ReceiverSettleMode,
-        ) -> Self {
+        ) -> &mut Self {
             self.options.receiver_settle_mode = Some(receiver_settle_mode);
             self
         }
-        #[allow(dead_code)]
-        pub fn with_source(mut self, source: impl Into<AmqpSource>) -> Self {
+
+        pub fn with_source(&mut self, source: impl Into<AmqpSource>) ->&mut Self {
             self.options.source = Some(source.into());
             self
         }
-        #[allow(dead_code)]
-        pub fn with_offered_capabilities(mut self, offered_capabilities: Vec<AmqpSymbol>) -> Self {
+        pub fn with_offered_capabilities(&mut self, offered_capabilities: Vec<AmqpSymbol>) ->&mut Self {
             self.options.offered_capabilities = Some(offered_capabilities);
             self
         }
         #[allow(dead_code)]
-        pub fn with_desired_capabilities(mut self, desired_capabilities: Vec<AmqpSymbol>) -> Self {
+        pub fn with_desired_capabilities(&mut self, desired_capabilities: Vec<AmqpSymbol>) -> &mut Self {
             self.options.desired_capabilities = Some(desired_capabilities);
             self
         }
-        #[allow(dead_code)]
+
         pub fn with_properties(
-            mut self,
+            &mut self,
             properties: impl Into<AmqpOrderedMap<AmqpSymbol, AmqpValue>>,
-        ) -> Self {
+        ) -> &mut Self {
             let properties_map: AmqpOrderedMap<AmqpSymbol, AmqpValue> =
                 properties.into().iter().collect();
 
             self.options.properties = Some(properties_map);
             self
         }
-        #[allow(dead_code)]
-        pub fn with_initial_delivery_count(mut self, initial_delivery_count: u32) -> Self {
+
+        pub fn with_initial_delivery_count(&mut self, initial_delivery_count: u32) -> &mut Self {
             self.options.initial_delivery_count = Some(initial_delivery_count);
             self
         }
-        pub fn with_max_message_size(mut self, max_message_size: u64) -> Self {
+
+        pub fn with_max_message_size(&mut self, max_message_size: u64) -> &mut Self {
             self.options.max_message_size = Some(max_message_size);
             self
         }
 
-        pub fn build(self) -> AmqpSenderOptions {
-            self.options
+        pub fn build(&mut self) -> AmqpSenderOptions {
+            self.options.clone()
```

Comment:
Just to confirm, this was the change related to removing the `Arc` we had discussed, right?

Reply:
No, this is related to the builder pattern changes to have a mutable builder.

The problem is that there's no easy way of expressing the immutable builder pattern via a flat C API - the mutable builder pattern works dramatically better for that case.

Reply:
I can play around with the immutable builder pattern, but it will require that each flat method return a newly boxed implementation of the builder (since each builder step returns a new instance of the builder, which then has to be boxed to return it to C clients).

An immutable builder has the advantage of mutating its internal state on each `with_xxx` call, which can be more easily expressed to C/C++ clients.

The amount of work I put into figuring out an immutable builder pattern depends on how important the immutable builder pattern is to our guidelines :).